### PR TITLE
refactor: improve full_chain_odd.py example

### PR DESCRIPTION
**Pull request by andiwand**
- align odd full chain with itk full chain

see #1513

### CODE REVIEW

File: Examples/Scripts/Python/full_chain_odd.py [hunk 1]
```diff
@@ -1,60 +1,78 @@
...
+field = acts.ConstantBField(acts.Vector3(0.0, 0.0, 2.0 * u.T))
+rnd = acts.examples.RandomNumbers(seed=42)
+
+s = acts.examples.Sequencer(events=100, numThreads=-1, logLevel=acts.logging.INFO)
+
+if not ttbar_pu200:
+    addParticleGun(
+        s,
+        MomentumConfig(1.0 * u.GeV, 10.0 * u.GeV, transverse=True),
+        EtaConfig(-4.0, 4.0, uniform=True),
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_odd.py:48]:
```suggestion
        EtaConfig(-3.0, 3.0, uniform=True),
```

File: Examples/Scripts/Python/full_chain_odd.py [hunk 2]
```diff
@@ -1,60 +1,78 @@
...
+        ),
+        rnd=rnd,
+        outputDirRoot=outputDir,
+    )
+
 addFatras(
     s,
     trackingGeometry,
     field,
+    ParticleSelectorConfig(eta=(-4.0, 4.0), pt=(150 * u.MeV, None), removeNeutral=True)
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_odd.py:69]:
```suggestion
    ParticleSelectorConfig(eta=(-3.0, 3.0), pt=(150 * u.MeV, None), removeNeutral=True)
```

File: Examples/Scripts/Python/full_chain_odd.py [hunk 3]
```diff
@@ -1,60 +1,78 @@
...
+        ParticleConfig(2, acts.PdgParticle.eMuon, randomizeCharge=True),
+        rnd=rnd,
+    )
+else:
+    addPythia8(
+        s,
+        hardProcess=["Top:qqbar2ttbar=on"],
+        npileup=200,
+        vtxGen=acts.examples.GaussianVertexGenerator(
+            stddev=acts.Vector4(0.0125 * u.mm, 0.0125 * u.mm, 55.5 * u.mm, 5.0 * u.ns),
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_odd.py:58]: This is the LHC beamspot and (maximum pile-up mu). I don't know what fantasy accelerator the ODD is for. Are these settings still OK?

File: Examples/Scripts/Python/full_chain_odd.py [hunk 4]
```diff
@@ -63,20 +81,26 @@
     outputDirRoot=outputDir,
     rnd=rnd,
 )
+
 addSeeding(
     s,
     trackingGeometry,
     field,
+    TruthSeedRanges(pt=(1.0 * u.GeV, None), eta=(-4.0, 4.0), nHits=(9, None))
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_odd.py:89]:
```suggestion
    TruthSeedRanges(pt=(1.0 * u.GeV, None), eta=(-3.0, 3.0), nHits=(9, None))
```
Also, how many hits should we normally expect for the ODD?

**Review by timadye — COMMENTED** [review]
Running with `ttbar_pu200 = True`, I got:
```
 PYTHIA Error in BeamRemnants::setKinematics: kinematics construction failed
```
Do we need to worry about that?

By the way, I noticed that this one has `addVertexFitting`. Do we want to include that in `full_chain_itk.py`, or is it still broken for the ITk?

File: Examples/Scripts/Python/full_chain_odd.py (hunk 3 above)
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:58]: I think this should be fine

File: Examples/Scripts/Python/full_chain_odd.py (hunk 4 above)
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:89]: hmm good point I will just revert to the default here. but the ODD has 9 layers if I am not mistaken

**Comment by timadye** [comment]
> did you see that in the ITk example too? I wonder why this would only show up in the ODD example

No, I didn't see it with the ITk example

**Comment by andiwand** [comment]
> By the way, I noticed that this one has addVertexFitting. Do we want to include that in full_chain_itk.py, or is it still broken for the ITk?

I think it should work with a similar setup like here after the fixes I added for the vertexing. since we do not have ambiguity resolution in Acts the results are quite funky sometimes but it should produce some OK results

**Comment by timadye** [comment]
> I think it should work with a similar setup like here after the fixes I added for the vertexing. since we do not have ambiguity resolution in Acts the results are quite funky sometimes but it should produce some OK results

Could you add that, so we have the example for the workshop? I can do it if you like, but you know more about this?

**Review by timadye — APPROVED** [review]

**Comment by andiwand** [comment]
@timadye I added the vertexing to the ITk example but didn't try it out yet because I don't know how to run it

I also need to look at the pythia error in the ODD example

**Comment by timadye** [comment]
> @timadye I added the vertexing to the ITk example but didn't try it out yet because I don't know how to run it

I can test it!

**Comment by andiwand** [comment]
> > @timadye I added the vertexing to the ITk example but didn't try it out yet because I don't know how to run it
> 
> I can test it!

thank you!

> Running with `ttbar_pu200 = True`, I got:
> 
> ```
>  PYTHIA Error in BeamRemnants::setKinematics: kinematics construction failed
> ```
> 
> Do we need to worry about that?

I did't see this error in the first few events. when did it happen for you? @timadye 

But I see a bunch of these

```
13:45:16    TrackFinding   ERROR     Step failed with EigenStepperError:3: Step size adjustment exceeds maximum trials
13:45:16    TrackFinding   ERROR     Propapation failed: EigenStepperError:3 Step size adjustment exceeds maximum trials with the initial parameters 72670 : 
 3.85578
 24.0004
 2.70553
0.682366
-1.54399
-11.8624
```

which is likely related to https://github.com/acts-project/acts/issues/1385 I think

**Comment by timadye** [comment]
> I did't see this error in the first few events. when did it happen for you? @timadye

It comes immediately after the `Sequencer` initialisation, before the first of the `EigenStepperError` errors:

```
13:02:22    Sequencer      INFO      Starting event loop with 1 threads
13:02:22    Sequencer      INFO        0 services
13:02:22    Sequencer      INFO        0 context decorators
13:02:22    Sequencer      INFO        1 readers
13:02:22    Sequencer      INFO        10 algorithms
13:02:22    Sequencer      INFO        12 writers
 PYTHIA Error in BeamRemnants::setKinematics: kinematics construction failed
13:02:27    TrackFinding   ERROR     Step failed with EigenStepperError:3: Step size adjustment exceeds maximum trials
```

I compiled with the latest key4hep installation (2022-08-22). I just tested again with your latest branch. The only changes are to set `ttbar_pu200 = True`. (I also reduced the number of events and tried with `numThreads=1`, but that doesn't affect the error.)

**Comment by andiwand** [comment]
very interesting I cannot reproduce it. it might be due to a different pythia version? I use `pythia8307`

also just freshly built acts from this branch

I will double check if my acts install was clean

File: Examples/Scripts/Python/full_chain_itk.py [hunk 5]
```diff
@@ -90,4 +93,25 @@
...
 )
 
+s.addAlgorithm(
+    acts.examples.TrackSelector(
+        level=acts.logging.INFO,
+        inputTrackParameters="fittedTrackParameters",
+        outputTrackParameters="trackparameters",
+        outputTrackIndices="outputTrackIndices",
+        removeNeutral=True,
+        absEtaMax=2.5,
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_itk.py:103]:
```suggestion
        absEtaMax=4.0,
```

**Review by timadye — CHANGES_REQUESTED** [review]
I suggested some changes for `full_chain_itk.py`, but I guess some of these could also apply to `full_chain_odd.py`.

File: Examples/Scripts/Python/full_chain_itk.py [hunk 6]
```diff
@@ -15,12 +15,15 @@
...
+    addVertexFitting,
+    VertexFinder,
 )
 
 ttbar_pu200 = False
 u = acts.UnitConstants
 geo_dir = pathlib.Path("acts-itk")
-outputDir = pathlib.Path.cwd()
+outputDir = pathlib.Path.cwd() / "itk_output"
+outputDir.mkdir(exist_ok=True)
```
[the remark this answers is not in the document]
Comment (timadye) [reply Examples/Scripts/Python/full_chain_itk.py:26]: sorry, just suggesting to remove:
```
outputDir.mkdir(exist_ok=True)
```

**Comment by timadye** [comment]
Since you move `outputDir` to `itk_output` subdirectory (a nice convenience), can use also specify it for the `Sequencer`?
```
s = acts.examples.Sequencer(events=100, numThreads=-1, outputDir=str(outputDir))
```
That puts the `timing.tsv` also in `itk_output`.

**Comment by timadye** [comment]
> @timadye I added the vertexing to the ITk example but didn't try it out yet because I don't know how to run it

with my suggested changes, it seems to run OK for dimuon events or `ttbar_pu200`. I don't know what sort of performance is to be expected with CKF output. For `ttbar_pu200`, I'm getting <`nRecoVtx`>=44 with <`nVtxReconstructable`>=148.

File: Examples/Scripts/Python/full_chain_itk.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_itk.py:26]: good point!

File: Examples/Scripts/Python/full_chain_itk.py [hunk 7]
```diff
@@ -34,7 +37,7 @@
         s,
         MomentumConfig(1.0 * u.GeV, 10.0 * u.GeV, transverse=True),
         EtaConfig(-4.0, 4.0, uniform=True),
-        ParticleConfig(1, acts.PdgParticle.eMuon, randomizeCharge=True),
+        ParticleConfig(2, acts.PdgParticle.eMuon, randomizeCharge=True),
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_itk.py:40]: yeah exactly with just one particle the vertexing will not do anything. for ODD 2 were enough so I didn't increase it further. but if one particle gets lots the vertex will be lost as well

File: Examples/Scripts/Python/full_chain_itk.py [hunk 8]
```diff
@@ -90,4 +93,25 @@
...
 
+s.addAlgorithm(
+    acts.examples.TrackSelector(
+        level=acts.logging.INFO,
+        inputTrackParameters="fittedTrackParameters",
+        outputTrackParameters="trackparameters",
+        outputTrackIndices="outputTrackIndices",
+        removeNeutral=True,
+        absEtaMax=2.5,
+        loc0Max=4.0 * u.mm,  # rho max
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_itk.py:104]: nice catch! hmmm I think I copied that from a different example without questioning it. lets see if we can get away with just dropping it

File: Examples/Scripts/Python/full_chain_itk.py [hunk 9]
```diff
@@ -90,4 +93,25 @@
     outputDirRoot=outputDir,
 )
 
+s.addAlgorithm(
+    acts.examples.TrackSelector(
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_itk.py:97]: makes sense! I also thought about that a while ago but forgot about it. I will follow up with a different PR

**Comment by andiwand** [comment]
> , outputDir=str(outputDir)

good idea! will add it

> with my suggested changes, it seems to run OK for dimuon events or `ttbar_pu200`. I don't know what sort of performance is to be expected with CKF output. For `ttbar_pu200`, I'm getting <`nRecoVtx`>=44 with <`nVtxReconstructable`>=148.

I am quite surprised that you can reconstruct 44 vertices. @paulgessinger ODD performed worse no? or was that just with AMVF?

File: Examples/Scripts/Python/full_chain_itk.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_itk.py:96]: https://github.com/acts-project/acts/pull/1545

File: Examples/Scripts/Python/full_chain_odd.py [hunk 10]
```diff
@@ -63,32 +80,38 @@
...
     acts.examples.TrackSelector(
         level=acts.logging.INFO,
         inputTrackParameters="fittedTrackParameters",
         outputTrackParameters="trackparameters",
         outputTrackIndices="outputTrackIndices",
         removeNeutral=True,
-        absEtaMax=2.5,
-        loc0Max=4.0 * u.mm,  # rho max
+        absEtaMax=3.0,
         ptMin=500 * u.MeV,
```
Comment (timadye) [inline Examples/Scripts/Python/full_chain_odd.py:111]: Sorry, there seems to be a bit of an inconsistency here. Above, we seem to be accepting that we can reconstruct above 1 GeV, but here we allow 0.5 GeV tracks in the vertices. If the vertex finder can accept lower pT tracks, then maybe we should do the same for the ITk (ie. undo one of my suggestions there).

File: Examples/Scripts/Python/full_chain_itk.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (timadye) [reply Examples/Scripts/Python/full_chain_itk.py:39]: I think you can resolve this comment...

File: Examples/Scripts/Python/full_chain_odd.py (hunk 10 above)
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:111]: no that makes a lot of sense I will change it

**Review by timadye — APPROVED** [review]
I don't know if I need to approve this again, but it does no harm I think.

The Pythia error message doesn't seem to stop the job working, so I think we can go ahead with this PR. If required to marge, maybe resolve that thread and when merged, unresolve it to continue the discussion?

**Comment by andiwand** [comment]
as far as I know you need a re-approve after additional changes are made.

> The Pythia error message doesn't seem to stop the job working, so I think we can go ahead with this PR. If required to marge, maybe resolve that thread and when merged, unresolve it to continue the discussion?

sounds good will do :+1:

**Comment by acts-project-service** [comment]
The backport to `develop/v19.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-develop/v19.x develop/v19.x
# Navigate to the new working tree
cd .worktrees/backport-develop/v19.x
# Create a new branch
git switch --create backport-1538-to-develop/v19.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick -x --mainline 1 e5d19700e5daa9d60922a2402fe61308eefb7639
# Push it to GitHub
git push --set-upstream origin backport-1538-to-develop/v19.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-develop/v19.x
```
Then, create a pull request where the `base` branch is `develop/v19.x` and the `compare`/`head` branch is `backport-1538-to-develop/v19.x`.

**Pull request merged**